pdagent/http.py

```diff
@@ -0,0 +1,93 @@
+# ("PSF"), and the Individual or Organization ("Licensee") accessing and
+# otherwise using this software ("Python") in source or binary form and
+# its associated documentation.
+#
+# 2. Subject to the terms and conditions of this License Agreement, PSF
+# hereby grants Licensee a nonexclusive, royalty-free, world-wide
+# license to reproduce, analyze, test, perform and/or display publicly,
+# prepare derivative works, distribute, and otherwise use Python
+# alone or in any derivative version, provided, however, that PSF's
+# License Agreement and PSF's notice of copyright, i.e., "Copyright (c)
+# 2001-2013 Python Software Foundation; All Rights Reserved" are retained in
+# Python alone or in any derivative version prepared by Licensee.
+#
+# 3. In the event Licensee prepares a derivative work that is based on
+# or incorporates Python or any part thereof, and wants to make
+# the derivative work available to others as provided herein, then
+# Licensee hereby agrees to include in any such work a brief summary of
+# the changes made to Python.
+#
+# 4. PSF is making Python available to Licensee on an "AS IS"
+# basis. PSF MAKES NO REPRESENTATIONS OR WARRANTIES, EXPRESS OR
+# IMPLIED. BY WAY OF EXAMPLE, BUT NOT LIMITATION, PSF MAKES NO AND
+# DISCLAIMS ANY REPRESENTATION OR WARRANTY OF MERCHANTABILITY OR FITNESS
+# FOR ANY PARTICULAR PURPOSE OR THAT THE USE OF PYTHON WILL NOT
+# INFRINGE ANY THIRD PARTY RIGHTS.
+#
+# 5. PSF SHALL NOT BE LIABLE TO LICENSEE OR ANY OTHER USERS OF PYTHON
+# FOR ANY INCIDENTAL, SPECIAL, OR CONSEQUENTIAL DAMAGES OR LOSS AS
+# A RESULT OF MODIFYING, DISTRIBUTING, OR OTHERWISE USING PYTHON,
+# OR ANY DERIVATIVE THEREOF, EVEN IF ADVISED OF THE POSSIBILITY THEREOF.
+#
+# 6. This License Agreement will automatically terminate upon a material
+# breach of its terms and conditions.
+#
+# 7. Nothing in this License Agreement shall be deemed to create any
+# relationship of agency, partnership, or joint venture between PSF and
+# Licensee. This License Agreement does not grant permission to use PSF
+# trademarks or trade name in a trademark sense to endorse or promote
+# products or services of Licensee, or any third party.
+#
+# 8. By copying, installing or otherwise using Python, Licensee
+# agrees to be bound by the terms and conditions of this License
+# Agreement.
+#
+
+import ssl
+
+from pdagent.pdagentutil import find_in_sys_path
+
+from six.moves.urllib import request
+
+DEFAULT_CA_CERTS_FILE = find_in_sys_path("pdagent/root_certs/ca_certs.pem")
+
+# For caching SSL contexts based on provided `ca_certs`.
+#
+# This imitates the previous caching behavior of this module. It isn't thread
+# safe, doesn't provide much optimization in its current form, so may be worth
+# removing or otherwise reevaluating.
+ssl_context_cache = dict()
+
+
+def urlopen(url, **kwargs):
+    ca_certs = kwargs.pop("ca_certs", DEFAULT_CA_CERTS_FILE)
+    context = kwargs.pop("context", _get_cached_context(ca_certs=ca_certs))
```

Comment:
Nit: you can use `kwargs.get` here instead, since you're not passing the `kwargs` to anything else afterwards. You could also do something like the following for the same reason:

    def urlopen(url, ca_certs=None, context=None):
        ca_certs = ca_certs or DEFAULT_CA_CERTS_FILES
        context = context or _get_cached_context(ca_certs=ca_certs))
        ... etc

Reply:
Might also be, you should be passing the additional arguments to `urlopen`:

    request.urlopen(url, context=context, **kwargs)

Reply:
Good feedback — this part is just a copy of how `httpswithverify` worked before. It’s something else I’d like to revisit in a subsequent PR.

Reply:
`httpswithverify` used to pass the rest of the keyword arguments through to the `open` function: 

https://github.com/PagerDuty/pdagent/pull/129/files/edd2d8a89e068240780fcece563a60aca857062b#diff-827a41294ecd5d77e62d265bf198ecddL173